### ISSUE TITLE
`ActiveRecord::Suppressor` not working on associations that are loaded via `inverse_of` on unsaved records

### ISSUE THREAD

**Issue by rosa** [rails/rails#51392]
### Steps to reproduce
This issue comes from [this comment](https://github.com/rails/rails/pull/50284#issuecomment-2004506200), and started happening after #50284. @rafaelfranca confirmed this behaviour (outlined by @andrewyoo) is intended:

```ruby
Loading development environment (Rails 7.2.0.alpha)
umami(dev)> u = User.new
=> 
#<User:0x000000012c399b30
...
umami(dev)> s = Shift.new user: u
=> #<Shift:0x0000000129e31d98 id: nil, user_id: nil>
umami(dev)> s.user
=> 
#<User:0x000000012c399b30>
umami(dev)> u.shifts
=> [#<Shift:0x0000000129e31d98 id: nil, user_id: nil>]
```

However, this has the side effect of breaking `ActiveRecord::Suppressor` if you happen to modify the association owner outside the suppression block. Here's a minimal reproduction script:

```ruby
# frozen_string_literal: true
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }
  gem "rails", github: "rails/rails", branch: "main"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"
require "debug"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)
ActiveRecord::Base.has_many_inversing = true

ActiveRecord::Schema.define do
  create_table :events, force: true do |t|
    t.string :action, limit: 255
    t.references :recording

    t.timestamps
  end

  create_table :recordings, force: true do |t|
    t.string :content

    t.timestamps
  end
end

class Recording < ActiveRecord::Base
  has_many :events, dependent: :destroy

  after_create :track_created

  def track_event(action)
    Event.create! recording: self, action: action
  end

  def copy
    transaction do
      recording = Event.suppress { Recording.create!(content: content) }
      recording.update!(updated_at: updated_at)
      recording.track_event(:copied_copy)
      track_event(:copied_original)
    end
  end

  private
    def track_created
      track_event :created
    end
end

class Event < ActiveRecord::Base
  belongs_to :recording, required: false
end

class BugTest < Minitest::Test
  def test_suppress_events_on_copies
    recording = Recording.create!(content: "original")
    assert_equal 1, Event.count

    recording.copy

    assert_equal 3, Event.count
  end
end
```

### Expected behavior
Unless it's decided that this is now the expected behaviour for suppression as well, and needs to be documented, I'd expect the above test to pass. 

I think, ideally, we'd fix `ActiveRecord::Suppressor` to deal with this case: a record that was set but not saved while the model was suppressed shouldn't be saved afterwards indirectly via the association owner. 

### Actual behavior
The test fails with 
```
Failure:
BugTest#test_suppress_events_on_copies [test.rb:75]:
Expected: 3
  Actual: 4
```
because we added a `created` event to the copy when we save the original recording outside the `Event.suppress` block. 

### System configuration
**Rails version**:  Rails version: `main` (7c68c52)
**Ruby version**: 3.3.0 (though it should happen with any version).

**Comment by casperisfine** [comment]
I think this is the desired behavior.

> ActiveRecord::Suppressor prevents the receiver from being saved during a given block.

Here the save happens outside of the block, so it seems legit. And even if it wasn't, I'm not even sure how we could possibly change that. We'd need to record some state on the unsaved record, and it would be very unclear when it would become "saveable" again.

**Comment by rosa** [comment]
Yes, that makes sense, but I think it's still a bit confusing, because with the code above, we're running essentially this inside the `suppress` block: 
```ruby
Event.create! recording: recording, action: action
```
And then, outside the `suppress` block, we're running this:
```ruby
recording.update(...)
```
So it feels quite disconnected that an event created within the suppress block via `Event.create(...)` and not, say, via `recording.events.create!(...)` suddenly gets saved outside the block. It seems to defeat the suppressor's purpose a bit, and I think the example documented in the `Suppressor` class, that goes like this https://github.com/rails/rails/blob/68b20b6513fe56ca80e4966628c231b4d6113bea/activerecord/lib/active_record/suppressor.rb#L9-L31

is very likely to hit this same issue if you opt for not disabling `inverse_of` in `has_many ..., :through` associations, which, at least apparently, is very much unrelated.

**Comment by casperisfine** [comment]
> suddenly gets saved outside the block

Unless I'm mistaken it get saves because `#update` is essentially a `#save` and `#save` also saves associated record. So it's not surprising to me at least.

>  It seems to defeat the suppressor's purpose a bit

IMO the suppressor API is a huge smell in the first place (but people are welcome to disagree).

**Comment by rosa** [comment]
> Unless I'm mistaken it get saves because `#update` is essentially a `#save` and `#save` also saves associated record. So it's not surprising to me at least.

Yeah, I didn't say it was surprising, but confusing and disconnected because we're saving an event created within the suppress block via `Event.create(...)` (in a callback as well) and not, say, via `recording.events.create!(...)`. In any case, I think it makes sense that this is the right behaviour, and perhaps this issue is enough in case someone runs in to this same problem in the future. 

> IMO the suppressor API is a huge smell in the first place (but people are welcome to disagree).

I don't have a strong opinion about this, TBH 😅 I do think the way it's implemented leaves a lot of room for confusing behaviour like this one, especially around associations.

**Comment by dhh** [comment]
Reviewed this with Rosa. I find this highly confusing. The kinda gotcha that's going to corrupt someone's system in production if they don't have some very fine-grained tests to catch it.

`Events.create!` should not be added itself to Recording's in-memory `has_many :events` automatically. If you want that, then you need to do `events.create!`. That to me is the root of this confusion. The suppression part is just a side effect of that root issue.

I'd consider this a blocker. It's a super subtle change, and the pay-off it might buy isn't worth it.
